jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DownloadObjectAction.kt

```diff
@@ -34,28 +41,114 @@ class DownloadObjectAction(private val project: Project, treeTable: S3TreeTable)
+    private fun getDownloadLocation(multipleFiles: Boolean): Path? {
         val baseDir = VfsUtil.getUserHomeDir()
-        val descriptor = FileSaverDescriptor(message("s3.download.object.action"), message("s3.download.object.description"))
-        val destination = FileChooserFactory.getInstance().createSaveFileDialog(descriptor, project).save(baseDir, file.name)?.file?.outputStream() ?: return
-        downloadAll(project, listOf(file.key to destination))
+
+        val descriptor = if (multipleFiles) {
+            FileChooserDescriptorFactory.createSingleFolderDescriptor().also {
+                it.description = message("s3.download.object.browse.description.multiple")
+            }
+        } else {
+            FileChooserDescriptorFactory.createSingleFileOrFolderDescriptor().also {
+                it.description = message("s3.download.object.browse.description.single")
+            }
+        }
+
+        descriptor.title = message("s3.download.object.browse.title")
+
+        // In order to prevent a confusing UX around when we can have folders and not, force the non native chooser
+        descriptor.isForcedToUseIdeaFileChooser = true
+
+        return FileChooser.chooseFile(descriptor, project, baseDir)?.path?.let {
+            Paths.get(it)
+        }
     }
 
-    private fun downloadAll(project: Project, files: List<Pair<String, OutputStream>>) {
+    private fun checkForConflicts(targetDirectory: Path, files: List<DownloadInfo>): List<DownloadInfo> {
+        val finalDownloads = mutableListOf<DownloadInfo>()
+
+        files.forEachIndexed { index, download ->
+            if (!download.diskLocation.exists()) {
+                finalDownloads.add(download)
+                return@forEachIndexed
+            }
+
+            val description = message(
+                "s3.download.object.conflict.description",
+                targetDirectory.relativize(download.diskLocation),
+                targetDirectory
+            )
+
+            val options = if (files.size == 1) {
+                arrayOf(
+                    message("s3.download.object.conflict.skip"),
+                    message("s3.download.object.conflict.overwrite")
+                )
+            } else {
+                arrayOf(
+                    message("s3.download.object.conflict.skip"),
+                    message("s3.download.object.conflict.overwrite"),
+                    message("s3.download.object.conflict.skip_rest"),
+                    message("s3.download.object.conflict.overwrite_rest")
+                )
+            }
+
+            val choice = Messages.showDialog(
+                project,
+                description,
+                message("s3.download.object.action"),
+                options,
+                0,
+                Messages.getQuestionIcon()
+            )
+
+            when (choice) {
+                0 -> LOG.info { "User requested skipping $download" }
```

Comment:
Can we do something about this relying on the order of the messages above?

Reply:
I thought enum, but we need the array to be in correct order and don't really want to rely on values() ordering.

ArrayList with "named" index may work and we sublist.toArray() based on needing 2 or 4...

Got any other ideas?

Reply:
Yeah the API seems to not be great, maybe all we need is a comment to not shuffle + build the second array out of the first (so they can't be different)?